### ISSUE TITLE
Bug/improvements in [z]imatcopy

### ISSUE THREAD

**Issue by klho** [xianyi/OpenBLAS#4038]
First, a bug: the amount of workspace memory allocated for the two-pass omatcopy block may not be enough, leading to bad memory writes. Here is a simple test case that demonstrates this:

```c
#include <stdio.h>
#include "cblas.h"

int main() {
    const int m = 4, n = 2, N = m*n;
    float a[N];
    for (int i = 0; i < N; ++i) a[i] = i;
    cblas_simatcopy(CblasRowMajor, CblasNoTrans, m, n, 1, a, n, n);
    return 0;
}
```

The cause of the problem is:

https://github.com/xianyi/OpenBLAS/blob/c74ee11376600ecf546240a78c6bf6a9728b91e0/interface/imatcopy.c#L152-L155

In this case, because `*lda == *ldb` (`= n`), the workspace memory `msize = n*n = 4` only.

Some other suggested improvements:

1. `msize` above may also be an over-allocation; only `(*rows) * (*cols)` entries are needed.
2. The NEW_IMATCOPY block does not require `*rows == *cols` if no transpose is being done (e.g., just in-place conjugation).

**Comment by martin-frbg** [comment]
Sorry for the late response, and thanks for looking into this. I have to admit that this function messed with my brain the last time I looked into it (when applying the fixes to the optimized in-place transform section).

**Comment by martin-frbg** [comment]
Closing as fixed by your PR #4039
